modules/ding_gatewayf/modules/ding_gatewayf_registration/ding_gatewayf_registration.module

```diff
@@ -0,0 +1,456 @@
+    '#attributes' => array(
+      // Auto-complete is disabled here as the registration may be preform on a
+      // public computer in the library and information about which user there
+      // is registered and the mail address is classified information.
+      'autocomplete' => 'off',
+    ),
+  );
+
+  // Get branch options form the active provider.
+  global $user;
+  $provider = _ding_provider_get_provider('reservation');
+  $options = ding_provider_invoke('reservation', 'options', $user);
+  $form['branch'] = $options[$provider['module'] . '_preferred_branch'];
+
+  $form['pin'] = array(
+    '#type' => 'password',
+    '#title' => t('Pin code'),
+    '#description' => t('Pin code with an max length of %length', array('%length' => ding_user_get_pincode_length())),
+    '#size' => 10,
+    '#maxlength' => ding_user_get_pincode_length(),
+    '#required' => TRUE,
+    '#attributes' => array(
+      'autocomplete' => 'off',
+    ),
+  );
+
+  $form['submit'] = array(
+    '#type' => 'submit',
+    '#value' => t('Create me'),
+  );
+
+  // Link cancel button to gatewayf logout to ensure that the user is signed
+  // out correctly.
+  $form['cancel'] = array(
+    '#markup' => l(t('Cancel'), DING_GATEWAYF_LOGOUT_URL, array('attributes' => array('class' => 'ding-gatewayf--cancel'))),
+  );
+
+  $form['#submit'] = array('ding_gatewayf_registration_acceptance_submit');
+
+  return $form;
+}
+
+/**
+ * Submit handler for the acceptance form.
+ *
+ * It tires to create the user at the provider and redirect the user.
+ */
+function ding_gatewayf_registration_acceptance_submit($form, $form_state) {
+  $values = $form_state['values'];
+
+  // This value was not store in the form to ensure that the user can reload the
+  // form.
+  $required_attributes = $_SESSION[DING_GATEWAYF_ATTRIBUTES];
+  unset($_SESSION[DING_GATEWAYF_ATTRIBUTES]);
+
+  // Check that the logged in user is over the age limit.
+  if (!_ding_gatewayf_registration_is_age_limit_met($required_attributes['schacPersonalUniqueID'])) {
+    $config = _ding_gatewayf_registration_load_config();
+    drupal_set_message(t('You are not allowed to automatically be created because of the age limit (!limit years). Please go to your local library to registry.', array('!limit' => $config['age_limit'])));
+
+    // Logout the user from gatewayf.
+    ding_gatewayf_logout();
+    return;
+  }
+
+  // Try to create the user at the provider.
+  try {
+    // Create user as the provider.
+    ding_provider_invoke('user', 'create', $required_attributes['schacPersonalUniqueID'], $values['pin'], $values['name'], $values['mail'], $values['branch']);
+
+    $_REQUEST['destination'] = DING_GATEWAYF_REGISTRATION_SUCCESS_URL;
+    _ding_gatewayf_provider_login($required_attributes);
+
+  }
+  catch (Exception $exception) {
+    drupal_set_message(t('You already have an user account at the library. So we have just logged you in.'));
+    _ding_gatewayf_provider_login($required_attributes);
```

Comment:
Does *any* exception thrown mean that the user already exist? That seems very suspicious.